### DISCUSSION TITLE
@vercel/node don't work with yarn workspaces

### DISCUSSION THREAD

**Discussion by eliectrlshift** [vercel/community#1524] [Help] [answered]
### Summary

Hello,

I'm currently having a problem using vercel and building via @vercel/node in a monorepo at the local dependency resolution level.

Here is the architecture of my monorepo:

![Capture d’écran 2023-02-14 151124](https://user-images.githubusercontent.com/116009764/218764034-990b5f77-d67e-4537-a266-2dc9b84728e2.png)

The two folders whose names have been deleted, including NextJS applications, these same applications are deployed via vercel without any problems in the monorepo, here the resolution of local packages works without problem.

I'm having a problem with the API.

Indeed I understood that I was bypassing the so-called classic build commands by using the "builds" part in the vercel.json file and its @vercel/node runtime.

Vercel project configuration :

![Capture d’écran 2023-02-14 150907](https://user-images.githubusercontent.com/116009764/218764270-0073de42-1fd7-4cb7-a5dc-8b85dabd944a.png)

Build logs : 

![Capture d’écran 2023-02-14 150929](https://user-images.githubusercontent.com/116009764/218764336-e90fe0fa-f163-420c-845e-ffd2a1547841.png)

API packages.json :

```
{
  "name": "api",
  "version": "0.0.0",
  ...
  "devDependencies": {
    ...
    "eslint-config-custom": "*",
    "prettier-config-custom": "*",
    "tsconfig-config-custom": "*",
    ...
  }
  ...
}
```

I try to references dependencies with file path instead of version like with `"eslint-config-custom": "file:../../packages/eslint-config-custom",` but the result is the same instead of the error message on vercel is maybe more explicit on what happens :

![image](https://user-images.githubusercontent.com/116009764/218765251-6377f3bd-1164-4cdd-b658-a451b6762234.png)

I understand that the definition of the root directory prevents access to files outside this one, but I have no problem at this level on the two other NextJS projects which use these same local packages and whose root directory is configured the same way.

What's more, in the sources of the build I can find the whole tree structure of my monorepo:

![image](https://user-images.githubusercontent.com/116009764/218766613-8906de20-5356-4847-a1fb-03fb20857ef8.png)

API vercel.json file :

```
{
  "version": 2,
  "env": {},
  "public": false,
  "github": {
    "enabled": false
  },
  "git": {
    "deploymentEnabled": {
      "main": true,
      "develop": true
    }
  },
  "builds": [
    {
      "src": "src/index.ts",
      "use": "@vercel/node"
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "src/index.ts",
      "methods": ["GET", "POST", "PUT", "DELETE"]
    }
  ]
}
```

I Hope someone could help me 🙏🏼 
Thank you !

**Comment by amyegan** [comment]
Thanks for reaching out, Elie! The [Node.js Runtime](https://vercel.com/docs/concepts/functions/serverless-functions/runtimes/node-js#) works with zero config from a project's `/api` directory. So you should be able to take advantage of a zero config build if you put the code into an `api` directory at the root of the API project. If you do need to customize the build, the [`functions`](https://vercel.com/docs/concepts/projects/project-configuration#functions) property is recommended over [`builds`](https://vercel.com/docs/concepts/projects/project-configuration#builds).

You may also be having some trouble related to accessing files outside of your serverless functions. The guide on [using files in Serverless Functions on Vercel](https://vercel.com/guides/how-can-i-use-files-in-serverless-functions) has more info on how to do that

**Reply by eliectrlshift** [reply]
Hi @amyegan, thank you for your answer.

I'm not sure I understand.

Using build and functions properties still requires the use of the @vercel/node runtime and my problem seems to come from resolving local packages when installing packages via @vercel/node.

I mean in any case no matter how, structure or language (js,ts) I provide my code to vercel, the latter will use the runtime to install the dependencies and I would be faced with my problem again right?

**Comment by eliectrlshift** [comment]
Could someone could help me on this ?

**Comment by Tawagot0** [comment]
I got the same problem

**Comment by amyegan** [comment, marked as answer]
There are too many possible causes of similar errors in monorepos for me to be able to pinpoint the exact cause here. But I believe that the following documentation is relevant and should help you get your site working:
- [Serverless Functions Overview](https://vercel.com/docs/concepts/functions/serverless-functions#)
- [Using the Node.js Runtime with Serverless Functions](https://vercel.com/docs/concepts/functions/serverless-functions/runtimes/node-js#)
- [Using Monorepos on Vercel](https://vercel.com/docs/concepts/monorepos#)
- [Deploying a Monorepo Using Yarn Workspaces to Vercel](https://vercel.com/guides/deploying-yarn-monorepos-to-vercel)
